monai/utils/module.py

```diff
@@ -193,7 +211,127 @@ def load_submodules(basemod, load_all: bool = True, exclude_pattern: str = "(.*[
     return submodules, err_mod
 
 
+class ComponentScanner:
+    """
+    Scan all the available classes and functions in the specified packages and modules.
+    Map the all the names and the module names in a table.
+
+    Args:
+        pkgs: the expected packages to scan modules and parse component names in the config.
+        modules: the expected modules in the packages to scan for all the components.
+            for example, to parser "LoadImage" in config, `pkgs` can be ["monai"], `modules` can be ["transforms"].
+        excludes: if any string of the `excludes` exists in the full module name, don't import this module.
+
+    """
+
+    def __init__(
+        self, pkgs: Sequence[str], modules: Optional[Sequence[str]] = None, excludes: Optional[Sequence[str]] = None
+    ):
+        self.pkgs = pkgs
+        self.modules = [] if modules is None else modules
+        self.excludes = [] if excludes is None else excludes
+        self._components_table = self._create_table()
+
+    def _create_table(self):
+        table: Dict[str, List] = {}
+        for pkg in self.pkgs:
+            package = import_module(pkg)
+
+            for _, modname, _ in walk_packages(path=package.__path__, prefix=package.__name__ + "."):
+                # if no modules specified, load all modules in the package
+                if all(s not in modname for s in self.excludes) and (
+                    len(self.modules) == 0 or any(name in modname for name in self.modules)
+                ):
+                    try:
+                        module = import_module(modname)
+                        for name, obj in inspect.getmembers(module):
+                            if (inspect.isclass(obj) or inspect.isfunction(obj)) and obj.__module__ == modname:
+                                if name not in table:
+                                    table[name] = []
+                                table[name].append(modname)
+                    except ModuleNotFoundError:
+                        pass
+        return table
+
+    def get_component_module_name(self, name):
+        """
+        Get the full module name of the class / function with specified name.
+        If target component name exists in multiple packages or modules, return all the paths.z
+
+        Args:
+            name: name of the expected class or function.
+
+        """
+        mods = self._components_table.get(name, None)
+        if isinstance(mods, list) and len(mods) == 1:
+            mods = mods[0]
+        return mods
+
+
+def locate(path: str):
```

Comment:
any idea why we don't use `pydoc.locate` directly?

Reply:
Hi @wyli ,

`pydoc.locate` can't work for several cases, that's why `Hydra` reimplemented it. Especially to import in **reverse order**:
https://github.com/facebookresearch/hydra/blob/main/hydra/_internal/utils.py#L558
For example, you can try `pydoc.locate` with this path "torch.optim._multi_tensor.adam.Adam":
https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/adam.py#L5
```
>>> import torch
>>> import pydoc
>>> 
>>> print(pydoc.locate("torch.optim._multi_tensor.adam.Adam"))
None
```

Thanks.

Reply:
I see, `torch.optim._multi_tensor` looks like a private module, perhaps we shouldn't support configuring with the packages private modules?

Reply:
I thought this is expected because
```py
>>> import torch
>>> torch.optim.Adam
<class 'torch.optim.adam.Adam'>
>>> torch.optim.adam.Adam
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'torch.optim' has no attribute 'adam'
>>> 
```

Reply:
Hi @wyli ,

But both below imports work:
```py
>>> from torch.optim import Adam
>>> 
>>> a = Adam(...)
>>> 
>>> from torch.optim.adam import Adam
>>> 
>>> b = Adam(...)
```

`torch.optim.Adam` should be supported to config and parse.
What do you think?

Thanks.

Reply:
agree that `torch.optim.Adam` should be supported,
this seems to work:
```py
>>> import pydoc
>>> pydoc.locate('torch.optim.Adam')
<class 'torch.optim.adam.Adam'>
>>> 
```
and the `torch.optim.Adam` usage is consistent with the documentation https://pytorch.org/docs/stable/generated/torch.optim.Adam.html#torch.optim.Adam

it's mentioned in this ticket: https://github.com/pytorch/pytorch/issues/271

so I feel it's good to use `pydoc` here.

Reply:
Hi @wyli ,

Thanks for your testing.
OK, as we aligned only need to support `<name>` for MONAI APIs, I want to double confirm that we don't have such "del XXX" case in any MONAI `__init__.py` source code, right?
If yes, let me remove this `locate` function and use `pydoc` directly.

Thanks.

Reply:
Hi @wyli @ericspod ,

BTW, I added `ComponentScanner` unit tests to ensure all the scanned MONAI components can support `locate`.
Do you have any other comments or concerns?

Thanks in advance.